proto/message.py

```diff
@@ -572,65 +606,64 @@ def __eq__(self, other):
         # Ask the other object.
         return NotImplemented
 
-    def __getattr__(self, key):
-        """Retrieve the given field's value.
-
-        In protocol buffers, the presence of a field on a message is
-        sufficient for it to always be "present".
-
-        For primitives, a value of the correct type will always be returned
-        (the "falsy" values in protocol buffers consistently match those
-        in Python). For repeated fields, the falsy value is always an empty
-        sequence.
-
-        For messages, protocol buffers does distinguish between an empty
-        message and absence, but this distinction is subtle and rarely
-        relevant. Therefore, this method always returns an empty message
-        (following the official implementation). To check for message
-        presence, use ``key in self`` (in other words, ``__contains__``).
-
-        .. note::
-
-            Some well-known protocol buffer types
-            (e.g. ``google.protobuf.Timestamp``) will be converted to
-            their Python equivalents. See the ``marshal`` module for
-            more details.
-        """
-        try:
-            pb_type = self._meta.fields[key].pb_type
-            pb_value = getattr(self._pb, key)
-            marshal = self._meta.marshal
-            return marshal.to_python(pb_type, pb_value, absent=key not in self)
-        except KeyError as ex:
-            raise AttributeError(str(ex))
-
     def __ne__(self, other):
         """Return True if the messages are unequal, False otherwise."""
         return not self == other
 
     def __repr__(self):
         return repr(self._pb)
 
-    def __setattr__(self, key, value):
-        """Set the value on the given field.
+    @property
+    def _pb(self):
+        if self._cached_pb is None:
+            self._cached_pb = self._meta.pb()
+        return self._cached_pb
+
+    @_pb.setter
+    def _pb(self, value):
+        self._cached_pb = value
+
+    def _mark_pb_stale(self, field_name: str):
+        """We often set fields on the proto-plus object (this), and do not immediately
+        sync them down to the underlying pb2 object. This mechanism tracks which
+        fields are stale in this way.
+        """
+        if not hasattr(self, '_stale_fields'):
+            self._stale_fields = set()
```

Comment:
Same here: I don't think we need to check `hasattr(self, '_stale_fields')`

Reply:
Same as above: our usage of `MyMessage.wrap()` bypasses the constructor 😲